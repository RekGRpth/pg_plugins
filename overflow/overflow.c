/*-------------------------------------------------------------------------
 *
 * overflow.c
 *		Common routines for overflow checks
 *
 * Copyright (c) 1996-2019, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		  overflow/overflow.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"
#include "fmgr.h"

#include "common/int.h"
#include "common/string.h"
#include "utils/builtins.h"

typedef enum
{
	NONE = 0,
	INT16,
	INT32,
	INT64,
	UINT16,
	UINT32,
	UINT64
} PGOverflowType;

typedef enum
{
	OPR_NONE = 0,
	OPR_ADD,
	OPR_SUB,
	OPR_MUL
} PGOverflowOpr;

/* smallint functions */
PG_FUNCTION_INFO_V1(pg_overflow_check);

Datum
pg_overflow_check(PG_FUNCTION_ARGS)
{
	int64		v1 = PG_GETARG_INT64(0);
	int64		v2 = PG_GETARG_INT64(1);
	int32		count = PG_GETARG_INT32(2);
	text	   *type_txt = PG_GETARG_TEXT_PP(3);
	text	   *opr_txt = PG_GETARG_TEXT_PP(4);
	bool		result = false;
	PGOverflowType type = NONE;
	PGOverflowOpr opr = OPR_NONE;
	int			i;
	char	   *type_str = text_to_cstring(type_txt);
	char	   *opr_str = text_to_cstring(opr_txt);

	/* extract variable type to work on */
	if (strcmp(type_str, "int16") == 0)
		type = INT16;
	else if (strcmp(type_str, "int32") == 0)
		type = INT32;
	else if (strcmp(type_str, "int64") == 0)
		type = INT32;
	else if (strcmp(type_str, "uint16") == 0)
		type = UINT16;
	else if (strcmp(type_str, "uint32") == 0)
		type = UINT32;
	else if (strcmp(type_str, "uint64") == 0)
		type = UINT64;
	else
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("unsupported overflow type")));

	/* extract variable type to work on */
	if (strcmp(opr_str, "add") == 0)
		opr = OPR_ADD;
	else if (strcmp(opr_str, "sub") == 0)
		opr = OPR_SUB;
	else if (strcmp(opr_str, "mul") == 0)
		opr = OPR_MUL;
	else
		ereport(ERROR,
				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
				 errmsg("unsupported overflow operation")));

	switch (type)
	{
		case INT16:
		{
			int16		val1 = (int16) v1;
			int16		val2 = (int16) v2;
			int16		val_res;

			switch (opr)
			{
				case OPR_ADD:
				{
					for (i = 0; i < count; i++)
						result = pg_add_s16_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_SUB:
				{
					for (i = 0; i < count; i++)
						result = pg_sub_s16_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_MUL:
				{
					for (i = 0; i < count; i++)
						result = pg_mul_s16_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_NONE:
					Assert(false);
					break;
			}
			break;
		}
		case INT32:
		{
			int32		val1 = (int32) v1;
			int32		val2 = (int32) v2;
			int32		val_res;

			switch (opr)
			{
				case OPR_ADD:
				{
					for (i = 0; i < count; i++)
						result = pg_add_s32_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_SUB:
				{
					for (i = 0; i < count; i++)
						result = pg_sub_s32_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_MUL:
				{
					for (i = 0; i < count; i++)
						result = pg_mul_s32_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_NONE:
					Assert(false);
					break;
			}
			break;
		}
		case INT64:
		{
			int64		val1 = (int64) v1;
			int64		val2 = (int64) v2;
			int64		val_res;

			switch (opr)
			{
				case OPR_ADD:
				{
					for (i = 0; i < count; i++)
						result = pg_add_s64_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_SUB:
				{
					for (i = 0; i < count; i++)
						result = pg_sub_s64_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_MUL:
				{
					for (i = 0; i < count; i++)
						result = pg_mul_s64_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_NONE:
					Assert(false);
					break;
			}
			break;
		}

		/* unsigned part */
		case UINT16:
		{
			uint16		val1 = (uint16) v1;
			uint16		val2 = (uint16) v2;
			uint16		val_res;

			switch (opr)
			{
				case OPR_ADD:
				{
					for (i = 0; i < count; i++)
						result = pg_add_u16_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_SUB:
				{
					for (i = 0; i < count; i++)
						result = pg_sub_u16_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_MUL:
				{
					for (i = 0; i < count; i++)
						result = pg_mul_u16_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_NONE:
					Assert(false);
					break;
			}
			break;
		}
		case UINT32:
		{
			uint32		val1 = (uint32) v1;
			uint32		val2 = (uint32) v2;
			uint32		val_res;

			switch (opr)
			{
				case OPR_ADD:
				{
					for (i = 0; i < count; i++)
						result = pg_add_u32_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_SUB:
				{
					for (i = 0; i < count; i++)
						result = pg_sub_u32_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_MUL:
				{
					for (i = 0; i < count; i++)
						result = pg_mul_u32_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_NONE:
					Assert(false);
					break;
			}
			break;
		}
		case UINT64:
		{
			uint64		val1 = (uint64) v1;
			uint64		val2 = (uint64) v2;
			uint64		val_res;

			switch (opr)
			{
				case OPR_ADD:
				{
					for (i = 0; i < count; i++)
						result = pg_add_u64_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_SUB:
				{
					for (i = 0; i < count; i++)
						result = pg_sub_u64_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_MUL:
				{
					for (i = 0; i < count; i++)
						result = pg_mul_u64_overflow(val1, val2, &val_res);
					break;
				}
				case OPR_NONE:
					Assert(false);
					break;
			}
			break;
		}

		case NONE:
			Assert(false);
			break;
	}

	PG_RETURN_BOOL(result);
}

/*
 * pg_strtouint64
 *	  Converts 'str' into an unsigned 64-bit integer.
 *
 * This has the identical API to strtoul(3), except that it will handle
 * 64-bit ints even where "long" is narrower than that.
 *
 * For the moment it seems sufficient to assume that the platform has
 * such a function somewhere; let's not roll our own.
 *
 * Note: this is an old implementation from PostgreSQL upstream which
 * has been removed in v13, used here for performance comparison.
 */
static uint64
pg_strtouint64_old(const char *str, char **endptr, int base)
{
#ifdef _MSC_VER                    /* MSVC only */
	return _strtoui64(str, endptr, base);
#elif defined(HAVE_STRTOULL) && SIZEOF_LONG < 8
	return strtoull(str, endptr, base);
#else
	return strtoul(str, endptr, base);
#endif
}


/*
 * Wrapper of pg_strtouint64_old used for performance comparison.
 */
PG_FUNCTION_INFO_V1(pg_strtouint64_old_check);
Datum
pg_strtouint64_old_check(PG_FUNCTION_ARGS)
{
	text	   *val_txt = PG_GETARG_TEXT_PP(0);
	int32		count = PG_GETARG_INT32(1);
	char	   *val = text_to_cstring(val_txt);
	uint64		res;
	int			i;

	for (i = 0; i < count; i++)
		res = pg_strtouint64_old(val, NULL, 10);

	PG_RETURN_INT64(res);
}

/*
 * Wrapper of pg_strtouint64 used for performance comparison.
 *
 * This ignores errors.
 */
PG_FUNCTION_INFO_V1(pg_strtouint64_new_check);
Datum
pg_strtouint64_new_check(PG_FUNCTION_ARGS)
{
	text	   *val_txt = PG_GETARG_TEXT_PP(0);
	int32		count = PG_GETARG_INT32(1);
	char	   *val = text_to_cstring(val_txt);
	uint64		res;
	int			i;

	/* just don't care about the status result here */
	for (i = 0; i < count; i++)
		(void) pg_strtouint64(val, &res);

	PG_RETURN_INT64(res);
}

/*
 * Wrapper of pg_strtoint() used for performance comparison.
 *
 * Note that this throws an ereport(ERROR) immediately on invalid
 * input or overflow.
 */
PG_FUNCTION_INFO_V1(pg_strtoint32_check_perf);
Datum
pg_strtoint32_check_perf(PG_FUNCTION_ARGS)
{
	text	   *val_txt = PG_GETARG_TEXT_PP(0);
	int32		count = PG_GETARG_INT32(1);
	char	   *val = text_to_cstring(val_txt);
	int32		res;
	int			i;

	for (i = 0; i < count; i++)
		res = pg_strtoint32_check(val);

	PG_RETURN_INT32(res);
}

/*
 * Wrapper of pg_strtoint32 used for performance comparison.
 *
 * This ignores errors.
 */
PG_FUNCTION_INFO_V1(pg_strtoint32_perf);
Datum
pg_strtoint32_perf(PG_FUNCTION_ARGS)
{
	text	   *val_txt = PG_GETARG_TEXT_PP(0);
	int32		count = PG_GETARG_INT32(1);
	char	   *val = text_to_cstring(val_txt);
	int32		res;
	int			i;

	/* just don't care about the status result here */
	for (i = 0; i < count; i++)
		(void) pg_strtoint32(val, &res);

	PG_RETURN_INT32(res);
}
