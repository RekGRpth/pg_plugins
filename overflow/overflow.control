# blackhole extension
comment = 'Overflow checks'
default_version = '1.0'
module_pathname = '$libdir/overflow'
relocatable = true
