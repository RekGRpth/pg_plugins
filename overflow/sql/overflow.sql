CREATE EXTENSION overflow;

-- smallint overflow checks
-- minimum int16 -32768
-- maximum int16 +32767
SELECT pg_add_int16_overflow(32767::smallint, 0::smallint);
SELECT pg_add_int16_overflow(32767::smallint, 1::smallint);
SELECT pg_sub_int16_overflow((-32768)::smallint, 0::smallint);
SELECT pg_sub_int16_overflow((-32768)::smallint, 1::smallint);
SELECT pg_mul_int16_overflow(32767::smallint, 1::smallint);
SELECT pg_mul_int16_overflow((-32768)::smallint, 1::smallint);
SELECT pg_mul_int16_overflow(32767::smallint, (-1)::smallint);
SELECT pg_mul_int16_overflow((-32768)::smallint, (-1)::smallint);

-- int overflow checks
-- minimum int32 -2147483648
-- maximum int32 +2147483647
SELECT pg_add_int32_overflow(2147483647, 0);
SELECT pg_add_int32_overflow(2147483647, 1);
SELECT pg_sub_int32_overflow(-2147483648, 0);
SELECT pg_sub_int32_overflow(-2147483648, 1);
SELECT pg_mul_int32_overflow(2147483647, 1);
SELECT pg_mul_int32_overflow(-2147483648, 1);
SELECT pg_mul_int32_overflow(2147483647, -1);
SELECT pg_mul_int32_overflow(-2147483648, -1);

-- bigint overflow checks
-- minimum int64 -9,223,372,036,854,775,808
-- maximum int64 +9,223,372,036,854,775,807
SELECT pg_add_int64_overflow(9223372036854775807::bigint, 0::bigint);
SELECT pg_add_int64_overflow(9223372036854775807::bigint, 1::bigint);
SELECT pg_sub_int64_overflow((-9223372036854775808)::bigint, 0::bigint);
SELECT pg_sub_int64_overflow((-9223372036854775808)::bigint, 1::bigint);
SELECT pg_mul_int64_overflow(9223372036854775807::bigint, 1::bigint);
SELECT pg_mul_int64_overflow((-9223372036854775808)::bigint, 1::bigint);
SELECT pg_mul_int64_overflow(9223372036854775807::bigint, (-1)::bigint);
SELECT pg_mul_int64_overflow((-9223372036854775808)::bigint, (-1)::bigint);

-- unsigned smallint overflow checks
-- maximum uint16 65535 => -1 as signed
SELECT pg_add_uint16_overflow((-1)::smallint, 0::smallint);
SELECT pg_add_uint16_overflow((-1)::smallint, 1::smallint);
SELECT pg_sub_uint16_overflow(0::smallint, 0::smallint);
SELECT pg_sub_uint16_overflow(0::smallint, 1::smallint);
SELECT pg_mul_uint16_overflow((-1)::smallint, 1::smallint);
SELECT pg_mul_uint16_overflow(0::smallint, 1::smallint);
SELECT pg_mul_uint16_overflow((-1)::smallint, (-1)::smallint);
SELECT pg_mul_uint16_overflow((0)::smallint, (-1)::smallint);

-- unsigned int overflow checks
-- maximum uint32 4294967295 => -1
SELECT pg_add_uint32_overflow(-1, 0);
SELECT pg_add_uint32_overflow(-1, 1);
SELECT pg_sub_uint32_overflow(0, 0);
SELECT pg_sub_uint32_overflow(0, 1);
SELECT pg_mul_uint32_overflow(-1, 1);
SELECT pg_mul_uint32_overflow(0, 1);
SELECT pg_mul_uint32_overflow(-1, -1);
SELECT pg_mul_uint32_overflow(0, -1);

-- unsigned bigint overflow checks
-- maximum uint64 18446744073709551615 => -1
SELECT pg_add_uint64_overflow((-1)::bigint, 0::bigint);
SELECT pg_add_uint64_overflow((-1)::bigint, 1::bigint);
SELECT pg_sub_uint64_overflow(0::bigint, 0::bigint);
SELECT pg_sub_uint64_overflow(0::bigint, 1::bigint);
SELECT pg_mul_uint64_overflow((-1)::bigint, 1::bigint);
SELECT pg_mul_uint64_overflow(0::bigint, 1::bigint);
SELECT pg_mul_uint64_overflow((-1)::bigint, (-1)::bigint);
SELECT pg_mul_uint64_overflow((0)::bigint, (-1)::bigint);
