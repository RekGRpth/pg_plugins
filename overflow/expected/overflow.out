CREATE EXTENSION overflow;
-- smallint overflow checks
-- minimum int16 -32768
-- maximum int16 +32767
SELECT pg_add_int16_overflow(32767::smallint, 0::smallint);
 pg_add_int16_overflow 
-----------------------
 f
(1 row)

SELECT pg_add_int16_overflow(32767::smallint, 1::smallint);
 pg_add_int16_overflow 
-----------------------
 t
(1 row)

SELECT pg_sub_int16_overflow((-32768)::smallint, 0::smallint);
 pg_sub_int16_overflow 
-----------------------
 f
(1 row)

SELECT pg_sub_int16_overflow((-32768)::smallint, 1::smallint);
 pg_sub_int16_overflow 
-----------------------
 t
(1 row)

SELECT pg_mul_int16_overflow(32767::smallint, 1::smallint);
 pg_mul_int16_overflow 
-----------------------
 f
(1 row)

SELECT pg_mul_int16_overflow((-32768)::smallint, 1::smallint);
 pg_mul_int16_overflow 
-----------------------
 f
(1 row)

SELECT pg_mul_int16_overflow(32767::smallint, (-1)::smallint);
 pg_mul_int16_overflow 
-----------------------
 f
(1 row)

SELECT pg_mul_int16_overflow((-32768)::smallint, (-1)::smallint);
 pg_mul_int16_overflow 
-----------------------
 t
(1 row)

-- int overflow checks
-- minimum int32 -2147483648
-- maximum int32 +2147483647
SELECT pg_add_int32_overflow(2147483647, 0);
 pg_add_int32_overflow 
-----------------------
 f
(1 row)

SELECT pg_add_int32_overflow(2147483647, 1);
 pg_add_int32_overflow 
-----------------------
 t
(1 row)

SELECT pg_sub_int32_overflow(-2147483648, 0);
 pg_sub_int32_overflow 
-----------------------
 f
(1 row)

SELECT pg_sub_int32_overflow(-2147483648, 1);
 pg_sub_int32_overflow 
-----------------------
 t
(1 row)

SELECT pg_mul_int32_overflow(2147483647, 1);
 pg_mul_int32_overflow 
-----------------------
 f
(1 row)

SELECT pg_mul_int32_overflow(-2147483648, 1);
 pg_mul_int32_overflow 
-----------------------
 f
(1 row)

SELECT pg_mul_int32_overflow(2147483647, -1);
 pg_mul_int32_overflow 
-----------------------
 f
(1 row)

SELECT pg_mul_int32_overflow(-2147483648, -1);
 pg_mul_int32_overflow 
-----------------------
 t
(1 row)

-- bigint overflow checks
-- minimum int64 -9,223,372,036,854,775,808
-- maximum int64 +9,223,372,036,854,775,807
SELECT pg_add_int64_overflow(9223372036854775807::bigint, 0::bigint);
 pg_add_int64_overflow 
-----------------------
 f
(1 row)

SELECT pg_add_int64_overflow(9223372036854775807::bigint, 1::bigint);
 pg_add_int64_overflow 
-----------------------
 t
(1 row)

SELECT pg_sub_int64_overflow((-9223372036854775808)::bigint, 0::bigint);
 pg_sub_int64_overflow 
-----------------------
 f
(1 row)

SELECT pg_sub_int64_overflow((-9223372036854775808)::bigint, 1::bigint);
 pg_sub_int64_overflow 
-----------------------
 t
(1 row)

SELECT pg_mul_int64_overflow(9223372036854775807::bigint, 1::bigint);
 pg_mul_int64_overflow 
-----------------------
 f
(1 row)

SELECT pg_mul_int64_overflow((-9223372036854775808)::bigint, 1::bigint);
 pg_mul_int64_overflow 
-----------------------
 f
(1 row)

SELECT pg_mul_int64_overflow(9223372036854775807::bigint, (-1)::bigint);
 pg_mul_int64_overflow 
-----------------------
 f
(1 row)

SELECT pg_mul_int64_overflow((-9223372036854775808)::bigint, (-1)::bigint);
 pg_mul_int64_overflow 
-----------------------
 t
(1 row)

